libs/platform/experience/carousel-navigation/carousel-navigation.component.ts

```diff
@@ -0,0 +1,410 @@
+   */
+  protected handleNext(e: Event): void {
+    let index;
+
+    if ((e as KeyboardEvent).altKey) {
+      index = this.items.length - 1;
+    } else {
+      const isRtl = window.getComputedStyle(this).direction === 'rtl';
+
+      const { left, width } = this.hostElement.getBoundingClientRect();
+      if (isRtl) {
+        index = this.items.findIndex((el) => {
+          return el.getBoundingClientRect().left <= left;
+        });
+      } else {
+        const x =
+          this.arrowNavigationBehavior === ArrowNavigationBehavior.Item
+            ? left + 10
+            : left + width;
+        index = this.items.findIndex(
+          (el) => el.getBoundingClientRect().left > x
+        );
+      }
+    }
+    this.scrollElementToIndex(index);
+  }
+
+  /**
+   * Handles the click on an indicator.
+   */
+  protected handleIndicatorClick(e: Event): void {
+    const target = e.target as HTMLInputElement;
+    this.scrollElementToIndex(parseInt(target.value));
+  }
+
+  /**
+   * Scrolls the carousel to the given index.
+   */
+  protected scrollElementToIndex(targetIndex: number): void {
+    const targetElement = this.items[targetIndex];
+    if (!targetElement) return;
+
+    const hostRect = this.hostElement.getBoundingClientRect();
+    const targetElementRect = targetElement.getBoundingClientRect();
+
+    const isRtl = window.getComputedStyle(this).direction === 'rtl';
+
+    const scrollPosition = isRtl
+      ? -(hostRect.right - targetElementRect.right)
+      : targetElementRect.left - hostRect.left;
+
+    const behavior =
+      this.scrollBehavior === CarouselScrollBehavior.Smooth ? 'smooth' : 'auto';
+
+    this.hostElement.scrollTo({
+      left: this.hostElement.scrollLeft + scrollPosition,
+      behavior,
+    });
+  }
+
+  protected get hostElement(): HTMLElement {
+    return this.getRootNode().host as HTMLElement;
+  }
+
+  /**
+   * Resolves all the carousel items. Carousel items can be rendered inside a composition
+   * or assigned to the main slot.
+   *
+   * The items can be listed inside a wrapper component, e.g. inside the product list component.
+   * In this case, the wrapper is supposed to have a 'display: contents' CSS property.
+   *
+   * Limitations: The composition layout supports individual order of components by setting the
+   * gridColumn (e.g., `gridColumn: 2`). This method will, however, not take this rendered order
+   * into account. There's currently no stable API to distinguish between the rendered order and
+   * the source order, but it will come: https://developer.chrome.com/blog/reading-order/.
+   *
+   * @returns An array of HTMLElements representing the child elements of the host component.
+   */
+  protected resolveItems(): HTMLElement[] {
+    const query =
+      '*:not(:is(oryx-carousel-navigation,slot,style,.indicators,button,oryx-button, input)';
+
+    const items = [
+      ...(this.hostElement?.shadowRoot?.querySelectorAll(query) ?? []),
```

Comment:
Are you just query all children in host's shadow dom? Can we narrow down the sample?

Reply:
simplified. Slotted children are as simple as the element.children. Assigned elements are only direct elements.